### PR TITLE
Added concurrency function in SDK

### PR DESCRIPTION
**Pull request by tyagiparth**

### CODE REVIEW

File: plugin/hydrate.go [hunk 1]
```diff
@@ -25,31 +31,59 @@ func (h *HydrateData) Clone() *HydrateData {
...
 // CanStart :: can this hydrate call - check whether all dependency hydrate functions have been completed
-func (h HydrateCall) CanStart(rowData *RowData) bool {
+func (h HydrateCall) CanStart(rowData *RowData, name string) bool {
 	for _, dep := range h.Depends {
 		if !helpers.StringSliceContains(rowData.getHydrateKeys(), dep) {
 			return false
 		}
 	}
-	return true
+	return concurrencyManager.StartIfAllowed(name, h.Config.MaxConcurrency)
```
[the remark this answers is not in the document]
Comment (tyagiparth) [reply plugin/hydrate.go:77]: yes, we are handling this here https://github.com/turbot/steampipe-plugin-sdk/pull/8/files#diff-51a868288d020cff3c5bb36badb8b720248be74cc791cf30fc7171086f949913R38

File: plugin/hydrate.go [hunk 2]
```diff
@@ -25,31 +31,59 @@ func (h *HydrateData) Clone() *HydrateData {
 type HydrateFunc func(context.Context, *QueryData, *HydrateData) (interface{}, error)
 
 // HydrateDependencies :: define the hydrate function dependencies - other hydrate functions which must be run first
+// DEPRECATED used HydrateConfig
 type HydrateDependencies struct {
 	Func    HydrateFunc
 	Depends []HydrateFunc
 }
 
+// HydrateDependencies :: define the hydrate function dependencies - other hydrate functions which must be run first
```
Comment (kaidaguerre) [inline plugin/hydrate.go:40]: change comment

File: plugin/hydrate.go [hunk 3]
```diff
@@ -25,31 +31,59 @@ func (h *HydrateData) Clone() *HydrateData {
...
+type HydrateConfig struct {
+	Func           HydrateFunc
+	MaxConcurrency int
+	// ConcurrencyMapKey ConcurrencyMapKeyFunc
+	//ShouldRetryError ErrorPredicate
+	//ShouldIgnoreError ErrorPredicate
+	Depends []HydrateFunc
+}
+
+type DefaultHydrateConfig struct {
```
Comment (kaidaguerre) [inline plugin/hydrate.go:50]: add a comment - explain this is used by the plugin

File: plugin/concurrency.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+type ConcurrencyManager struct {
+	mut        sync.Mutex
+	totalCalls int
+	// map of max calls allowed for each func
+	callMaxMap map[string]int
+	// map of current calls for each func
+	callMap map[string]int
+}
+
+func newConcurrencyManager() *ConcurrencyManager {
```
Comment (kaidaguerre) [inline plugin/concurrency.go:20]: for now remove (comment out) `defaultMaxPerCall` - our config schema does not include it

We need to move defaultMaxCalls into the concurrency manager and populate from the plugin  `DefaultHydrateConfig.MaxConcurrency` (add same var to `ConcurrencyManmager` )

callMaxMap either needs to be populated when we start concurrency manager from the hydrate configs of the required hydrate functions, or removed and use the hydrate config directly

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

**Review by kaidaguerre — APPROVED** [review]

**Pull request merged**